Nurx Assignment to predict churn rate, variables that impact churn rate and recommendations to improve churn rate. 

In [1]:
#Import libraries
import numpy  as np 
import pandas as pd 
import os
import gc
import matplotlib.pyplot as plt
import seaborn as sns

In [5]:
#Import dataset
data = pd.read_csv("Telco-Customer-Churn.csv")

In [4]:
print(data)

       customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0             NaN     NaN            NaN     NaN        NaN     NaN   
1      7590-VHVEG  Female            0.0     Yes         No     1.0   
2             NaN     NaN            NaN     NaN        NaN     NaN   
3      5575-GNVDE    Male            0.0      No         No    34.0   
4             NaN     NaN            NaN     NaN        NaN     NaN   
5      3668-QPYBK    Male            0.0      No         No     2.0   
6             NaN     NaN            NaN     NaN        NaN     NaN   
7      7795-CFOCW    Male            0.0      No         No    45.0   
8             NaN     NaN            NaN     NaN        NaN     NaN   
9      9237-HQITU  Female            0.0      No         No     2.0   
10            NaN     NaN            NaN     NaN        NaN     NaN   
11     9305-CDSKC  Female            0.0      No         No     8.0   
12            NaN     NaN            NaN     NaN        NaN     NaN   
13    

In [6]:
#Look for rows and columns in the dataset
data.shape

(14086, 21)

In [7]:
#Look for first 5 rows
data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0         NaN     NaN            NaN     NaN        NaN     NaN          NaN   
1  7590-VHVEG  Female            0.0     Yes         No     1.0           No   
2         NaN     NaN            NaN     NaN        NaN     NaN          NaN   
3  5575-GNVDE    Male            0.0      No         No    34.0          Yes   
4         NaN     NaN            NaN     NaN        NaN     NaN          NaN   

      MultipleLines InternetService OnlineSecurity  ...  DeviceProtection  \
0               NaN             NaN            NaN  ...               NaN   
1  No phone service             DSL             No  ...                No   
2               NaN             NaN            NaN  ...               NaN   
3                No             DSL            Yes  ...               Yes   
4               NaN             NaN            NaN  ...               NaN   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0         NaN         NaN             NaN             NaN              NaN   
1          No          No              No  Month-to-month              Yes   
2         NaN         NaN             NaN             NaN              NaN   
3          No          No              No        One year               No   
4         NaN         NaN             NaN             NaN              NaN   

      PaymentMethod MonthlyCharges  TotalCharges Churn  
0               NaN            NaN           NaN   NaN  
1  Electronic check          29.85         29.85    No  
2               NaN            NaN           NaN   NaN  
3      Mailed check          56.95        1889.5    No  
4               NaN            NaN           NaN   NaN  

[5 rows x 21 columns]

In [8]:
#Look for last 5 rows
data.tail()

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
14081  4801-JZAZL  Female            0.0     Yes        Yes    11.0   
14082         NaN     NaN            NaN     NaN        NaN     NaN   
14083  8361-LTMKD    Male            1.0     Yes         No     4.0   
14084         NaN     NaN            NaN     NaN        NaN     NaN   
14085  3186-AJIEK    Male            0.0      No         No    66.0   

      PhoneService     MultipleLines InternetService OnlineSecurity  ...   \
14081           No  No phone service             DSL            Yes  ...    
14082          NaN               NaN             NaN            NaN  ...    
14083          Yes               Yes     Fiber optic             No  ...    
14084          NaN               NaN             NaN            NaN  ...    
14085          Yes                No     Fiber optic            Yes  ...    

      DeviceProtection TechSupport StreamingTV StreamingMovies  \
14081               No          No          No              No   
14082              NaN         NaN         NaN             NaN   
14083               No          No          No              No   
14084              NaN         NaN         NaN             NaN   
14085              Yes         Yes         Yes             Yes   

             Contract PaperlessBilling              PaymentMethod  \
14081  Month-to-month              Yes           Electronic check   
14082             NaN              NaN                        NaN   
14083  Month-to-month              Yes               Mailed check   
14084             NaN              NaN                        NaN   
14085        Two year              Yes  Bank transfer (automatic)   

      MonthlyCharges  TotalCharges Churn  
14081          29.60        346.45    No  
14082            NaN           NaN   NaN  
14083          74.40         306.6   Yes  
14084            NaN           NaN   NaN  
14085         105.65        6844.5    No  

[5 rows x 21 columns]

In [9]:
data.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14086 entries, 0 to 14085
Data columns (total 21 columns):
customerID          7043 non-null object
gender              7043 non-null object
SeniorCitizen       7043 non-null float64
Partner             7043 non-null object
Dependents          7043 non-null object
tenure              7043 non-null float64
PhoneService        7043 non-null object
MultipleLines       7043 non-null object
InternetService     7043 non-null object
OnlineSecurity      7043 non-null object
OnlineBackup        7043 non-null object
DeviceProtection    7043 non-null object
TechSupport         7043 non-null object
StreamingTV         7043 non-null object
StreamingMovies     7043 non-null object
Contract            7043 non-null object
PaperlessBilling    7043 non-null object
PaymentMethod       7043 non-null object
MonthlyCharges      7043 non-null float64
TotalCharges        7043 non-null object
Churn               7043 non-null object
dtypes: float64(3), object(

In [12]:
#Check for duplicates
data.duplicated()

0        False
1        False
2         True
3        False
4         True
5        False
6         True
7        False
8         True
9        False
10        True
11       False
12        True
13       False
14        True
15       False
16        True
17       False
18        True
19       False
20        True
21       False
22        True
23       False
24        True
25       False
26        True
27       False
28        True
29       False
         ...  
14056     True
14057    False
14058     True
14059    False
14060     True
14061    False
14062     True
14063    False
14064     True
14065    False
14066     True
14067    False
14068     True
14069    False
14070     True
14071    False
14072     True
14073    False
14074     True
14075    False
14076     True
14077    False
14078     True
14079    False
14080     True
14081    False
14082     True
14083    False
14084     True
14085    False
Length: 14086, dtype: bool

In [19]:
#Drop Duplicates from dataset
data1= data.drop_duplicates()

In [18]:
#Look at the dataset after removing dups
data1.shape

(7044, 21)

In [20]:
#Look for missing values
data1.isnull().values.any()

True

In [21]:
#Drop rows where all elements are missing
data1.dropna(how='all')

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
1      7590-VHVEG  Female            0.0     Yes         No     1.0   
3      5575-GNVDE    Male            0.0      No         No    34.0   
5      3668-QPYBK    Male            0.0      No         No     2.0   
7      7795-CFOCW    Male            0.0      No         No    45.0   
9      9237-HQITU  Female            0.0      No         No     2.0   
11     9305-CDSKC  Female            0.0      No         No     8.0   
13     1452-KIOVK    Male            0.0      No        Yes    22.0   
15     6713-OKOMC  Female            0.0      No         No    10.0   
17     7892-POOKP  Female            0.0     Yes         No    28.0   
19     6388-TABGU    Male            0.0      No        Yes    62.0   
21     9763-GRSKD    Male            0.0     Yes        Yes    13.0   
23     7469-LKBCI    Male            0.0      No         No    16.0   
25     8091-TTVAX    Male            0.0     Yes         No    58.0   
27     0280-XJGEX    Male            0.0      No         No    49.0   
29     5129-JLPIS    Male            0.0      No         No    25.0   
31     3655-SNQYZ  Female            0.0     Yes        Yes    69.0   
33     8191-XWSZG  Female            0.0      No         No    52.0   
35     9959-WOFKT    Male            0.0      No        Yes    71.0   
37     4190-MFLUW  Female            0.0     Yes        Yes    10.0   
39     4183-MYFRB  Female            0.0      No         No    21.0   
41     8779-QRDMV    Male            1.0      No         No     1.0   
43     1680-VDCWW    Male            0.0     Yes         No    12.0   
45     1066-JKSGK    Male            0.0      No         No     1.0   
47     3638-WEABW  Female            0.0     Yes         No    58.0   
49     6322-HRPFA    Male            0.0     Yes        Yes    49.0   
51     6865-JZNKO  Female            0.0      No         No    30.0   
53     6467-CHFZW    Male            0.0     Yes        Yes    47.0   
55     8665-UTDHZ    Male            0.0     Yes        Yes     1.0   
57     5248-YGIJN    Male            0.0     Yes         No    72.0   
59     8773-HHUOZ  Female            0.0      No        Yes    17.0   
...           ...     ...            ...     ...        ...     ...   
14027  1685-BQULA  Female            0.0      No         No    40.0   
14029  9053-EJUNL    Male            0.0      No         No    41.0   
14031  0666-UXTJO    Male            1.0     Yes         No    34.0   
14033  1471-GIQKQ  Female            0.0      No         No     1.0   
14035  4807-IZYOZ  Female            0.0      No         No    51.0   
14037  1122-JWTJW    Male            0.0     Yes        Yes     1.0   
14039  9710-NJERN  Female            0.0      No         No    39.0   
14041  9837-FWLCH    Male            0.0     Yes        Yes    12.0   
14043  1699-HPSBG    Male            0.0      No         No    12.0   
14045  7203-OYKCT    Male            0.0      No         No    72.0   
14047  1035-IPQPU  Female            1.0     Yes         No    63.0   
14049  7398-LXGYX    Male            0.0     Yes         No    44.0   
14051  2823-LKABH  Female            0.0      No         No    18.0   
14053  8775-CEBBJ  Female            0.0      No         No     9.0   
14055  0550-DCXLH    Male            0.0      No         No    13.0   
14057  9281-CEDRU  Female            0.0     Yes         No    68.0   
14059  2235-DWLJU  Female            1.0      No         No     6.0   
14061  0871-OPBXW  Female            0.0      No         No     2.0   
14063  3605-JISKB    Male            1.0     Yes         No    55.0   
14065  6894-LFHLY    Male            1.0      No         No     1.0   
14067  9767-FFLEM    Male            0.0      No         No    38.0   
14069  0639-TSIQW  Female            0.0      No         No    67.0   
14071  8456-QDAVC    Male            0.0      No         No    19.0   
14073  7750-EYXWZ  Female            0.0      No         No    12.0   
14075  2569-WGERO  Female      

In [29]:
data1.shape

(7044, 21)

In [38]:
#Drop rows with missing values of customerID
data2 = data1.dropna(subset=['customerID', 'Churn'])

In [39]:
data2.shape

(7043, 21)

In [61]:
#Look at distribution of gender data
gender=data2.groupby("gender").size()
gender

gender
Female    3488
Male      3555
dtype: int64

In [62]:
#Look at distribution of Senior Citizen data 
SeniorCitizen=data2.groupby("SeniorCitizen").size()
SeniorCitizen

SeniorCitizen
0.0    5901
1.0    1142
dtype: int64

In [63]:
#Look at distribution of Partner data 
Partner=data2.groupby("Partner").size()
Partner

Partner
No     3641
Yes    3402
dtype: int64

In [64]:
#Look at distribution of Dependents data 
Dependents=data2.groupby("Dependents").size()
Dependents

Dependents
No     4933
Yes    2110
dtype: int64

In [45]:
#Look at distribution of tenure data 
tenure=data.groupby("tenure").size()
tenure

tenure
0.0      11
1.0     613
2.0     238
3.0     200
4.0     176
5.0     133
6.0     110
7.0     131
8.0     123
9.0     119
10.0    116
11.0     99
12.0    117
13.0    109
14.0     76
15.0     99
16.0     80
17.0     87
18.0     97
19.0     73
20.0     71
21.0     63
22.0     90
23.0     85
24.0     94
25.0     79
26.0     79
27.0     72
28.0     57
29.0     72
       ... 
43.0     65
44.0     51
45.0     61
46.0     74
47.0     68
48.0     64
49.0     66
50.0     68
51.0     68
52.0     80
53.0     70
54.0     68
55.0     64
56.0     80
57.0     65
58.0     67
59.0     60
60.0     76
61.0     76
62.0     70
63.0     72
64.0     80
65.0     76
66.0     89
67.0     98
68.0    100
69.0     95
70.0    119
71.0    170
72.0    362
Length: 73, dtype: int64

In [65]:
#Look at distribution of PhoneService data 
PhoneService=data2.groupby("PhoneService").size()
PhoneService

PhoneService
No      682
Yes    6361
dtype: int64

In [66]:
#Look at distribution of PhoneService data 
MultipleLines=data2.groupby("MultipleLines").size()
MultipleLines 

MultipleLines
No                  3390
No phone service     682
Yes                 2971
dtype: int64

In [67]:
#Look at distribution of InternetService data 
InternetService=data2.groupby("InternetService").size()
InternetService 

InternetService
DSL            2421
Fiber optic    3096
No             1526
dtype: int64

In [68]:
#Look at distribution of Online Security data 
OnlineSecurity=data2.groupby("OnlineSecurity").size()
OnlineSecurity

OnlineSecurity
No                     3498
No internet service    1526
Yes                    2019
dtype: int64

In [69]:
#Look at distribution of Online Security data 
DeviceProtection=data2.groupby("DeviceProtection").size()
DeviceProtection

DeviceProtection
No                     3095
No internet service    1526
Yes                    2422
dtype: int64

In [70]:
#Look at distribution of Tech Support data 
TechSupport=data2.groupby("TechSupport").size()
TechSupport

TechSupport
No                     3473
No internet service    1526
Yes                    2044
dtype: int64

In [71]:
#Look at distribution of Streaming TV data 
StreamingTV=data2.groupby("StreamingTV").size()
StreamingTV

StreamingTV
No                     2810
No internet service    1526
Yes                    2707
dtype: int64

In [72]:
#Look at distribution of Streaming Movies data 
StreamingMovies=data2.groupby("StreamingMovies").size()
StreamingMovies

StreamingMovies
No                     2785
No internet service    1526
Yes                    2732
dtype: int64

In [73]:
#Look at distribution of Contract  data 
Contract = data2.groupby("Contract").size()
Contract 

Contract
Month-to-month    3875
One year          1473
Two year          1695
dtype: int64

In [74]:
#Look at distribution of PaperlessBilling data 
PaperlessBilling=data2.groupby("PaperlessBilling").size()
PaperlessBilling

PaperlessBilling
No     2872
Yes    4171
dtype: int64

In [75]:
#Look at distribution of PaymentMethod data 
PaymentMethod=data2.groupby("PaymentMethod").size()
PaymentMethod

PaymentMethod
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Electronic check             2365
Mailed check                 1612
dtype: int64

In [59]:
#Look at distribution of MonthlyCharges data 
MonthlyCharges=data.groupby("MonthlyCharges").size()
MonthlyCharges

MonthlyCharges
18.25      1
18.40      1
18.55      1
18.70      2
18.75      1
18.80      7
18.85      5
18.90      2
18.95      6
19.00      7
19.05      9
19.10     12
19.15     11
19.20     17
19.25     18
19.30     24
19.35     25
19.40     31
19.45     28
19.50     32
19.55     40
19.60     37
19.65     43
19.70     43
19.75     39
19.80     38
19.85     45
19.90     44
19.95     44
20.00     43
          ..
115.80     4
115.85     1
116.00     1
116.05     3
116.10     2
116.15     1
116.20     1
116.25     3
116.30     1
116.40     1
116.45     1
116.50     1
116.55     1
116.60     2
116.75     1
116.80     1
116.85     1
116.95     1
117.15     1
117.20     1
117.35     1
117.45     1
117.50     1
117.60     1
117.80     1
118.20     1
118.35     1
118.60     2
118.65     1
118.75     1
Length: 1585, dtype: int64

In [60]:
#Look at distribution of TotalCharges data 
TotalCharges=data.groupby("TotalCharges").size()
TotalCharges

TotalCharges
           11
100.2       1
100.25      1
100.35      1
100.4       1
100.8       1
100.9       1
1001.2      1
1001.5      1
1003.05     1
1004.35     1
1004.5      1
1004.75     1
1005.7      1
1006.9      1
1007.8      1
1007.9      1
1008.55     1
1008.7      2
1009.25     1
101.1       1
101.45      1
101.65      1
101.9       1
1010        1
1011.05     1
1011.5      1
1011.8      1
1012.4      1
1013.05     1
           ..
98.5        2
980.35      1
981.45      1
982.95      1
985.05      1
987.95      2
989.05      2
989.45      1
99.45       1
99.6        1
99.75       2
99.95       1
990.3       1
990.45      1
990.85      1
990.9       1
992.7       1
993.15      1
994.55      1
994.8       1
995.35      1
996.45      1
996.85      1
996.95      1
997.65      1
997.75      1
998.1       1
999.45      1
999.8       1
999.9       1
Length: 6531, dtype: int64

In [78]:
Churn=data.groupby('Churn').size()
Churn

Churn
No     5174
Yes    1869
dtype: int64

In [80]:
#Gender impact on Churn?
Gender_Churn=data2.groupby(['gender', 'Churn']).size()
Gender_Churn

gender  Churn
Female  No       2549
        Yes       939
Male    No       2625
        Yes       930
dtype: int64

Gender doesn't have any impact on churn rate. The sample size is equally distributed with male and female. Male and female both have equal likelihood of churn (50% churn rate for Males and Females). 

In [81]:
SeniorCitizen_Churn=data2.groupby(['SeniorCitizen', 'Churn']).size()
SeniorCitizen_Churn

SeniorCitizen  Churn
0.0            No       4508
               Yes      1393
1.0            No        666
               Yes       476
dtype: int64

It seems SeniorCitizen have low churn rate as compared to non-senior citizens. 

In [82]:
Partner_Churn=data2.groupby(['Partner', 'Churn']).size()
Partner_Churn

Partner  Churn
No       No       2441
         Yes      1200
Yes      No       2733
         Yes       669
dtype: int64

Partner doesn't have impact on churn rate. Having partner or not doesn't increase or decrease churn rate. 

In [83]:
Dependents_Churn=data2.groupby(['Dependents', 'Churn']).size()
Dependents_Churn

Dependents  Churn
No          No       3390
            Yes      1543
Yes         No       1784
            Yes       326
dtype: int64

Dependents variable impacts Churn rate. Having dependents decreases churn rate. 

In [84]:
tenure_Churn=data2.groupby(['tenure', 'Churn']).size()
tenure_Churn

tenure  Churn
0.0     No        11
1.0     No       233
        Yes      380
2.0     No       115
        Yes      123
3.0     No       106
        Yes       94
4.0     No        93
        Yes       83
5.0     No        69
        Yes       64
6.0     No        70
        Yes       40
7.0     No        80
        Yes       51
8.0     No        81
        Yes       42
9.0     No        73
        Yes       46
10.0    No        71
        Yes       45
11.0    No        68
        Yes       31
12.0    No        79
        Yes       38
13.0    No        71
        Yes       38
14.0    No        52
        Yes       24
15.0    No        62
                ... 
58.0    No        56
        Yes       11
59.0    No        52
        Yes        8
60.0    No        70
        Yes        6
61.0    No        68
        Yes        8
62.0    No        65
        Yes        5
63.0    No        68
        Yes        4
64.0    No        76
        Yes        4
65.0    No        67
        Yes        9

Tenure is a very important variable in predicting churn. The smaller the tenure, the higher the churn. The longer the tenure period, the smaller churn rate. 

In [379]:
PhoneService_Churn=data2.groupby(['PhoneService', 'Churn']).size()
PhoneService_Churn

PhoneService  Churn
0             0         512
              1         170
1             0        4662
              1        1699
dtype: int64

Phone Service doesn't influence churn rate. Irrelevent of having phone service or not, there is no impact on churn. 

In [380]:
MultipleLines_Churn=data2.groupby(['MultipleLines', 'Churn']).size()
MultipleLines_Churn

MultipleLines  Churn
0              0        2541
               1         849
1              0        2121
               1         850
2              0         512
               1         170
dtype: int64

Having Multiple Lines or not has no impact on Churn Rate. 

In [87]:
InternetService_Churn=data2.groupby(['InternetService', 'Churn']).size()
InternetService_Churn

InternetService  Churn
DSL              No       1962
                 Yes       459
Fiber optic      No       1799
                 Yes      1297
No               No       1413
                 Yes       113
dtype: int64

In [88]:
OnlineSecurity_Churn=data2.groupby(['OnlineSecurity', 'Churn']).size()
OnlineSecurity_Churn

OnlineSecurity       Churn
No                   No       2037
                     Yes      1461
No internet service  No       1413
                     Yes       113
Yes                  No       1724
                     Yes       295
dtype: int64

People with no online security are more likely to churn than people with online security. People with  No Online security have 78% churn rate while people with online security have 16% churn rate. 

In [89]:
TechSupport_Churn=data2.groupby(['TechSupport', 'Churn']).size()
TechSupport_Churn

TechSupport          Churn
No                   No       2027
                     Yes      1446
No internet service  No       1413
                     Yes       113
Yes                  No       1734
                     Yes       310
dtype: int64

People with no tech support are more likely to churn than people with tech support. People with No tech Support have 77% churn rate while people with tech support have 17% churn rate. 

In [91]:
StreamingTV_Churn=data2.groupby(['StreamingTV', 'Churn']).size()
StreamingTV_Churn

StreamingTV          Churn
No                   No       1868
                     Yes       942
No internet service  No       1413
                     Yes       113
Yes                  No       1893
                     Yes       814
dtype: int64

StreamingTV or StreamingMovies have minimal impact on Churn Rate. 

In [90]:
StreamingMovies_Churn=data2.groupby(['StreamingMovies', 'Churn']).size()
StreamingMovies_Churn

StreamingMovies      Churn
No                   No       1847
                     Yes       938
No internet service  No       1413
                     Yes       113
Yes                  No       1914
                     Yes       818
dtype: int64

In [92]:
Contract_Churn=data2.groupby(['Contract', 'Churn']).size()
Contract_Churn

Contract        Churn
Month-to-month  No       2220
                Yes      1655
One year        No       1307
                Yes       166
Two year        No       1647
                Yes        48
dtype: int64

 People with month-to-month contract duration have higher churn rate than people with two year contract period. People with month-to-month contract duration have 89% churn rate as compared to 1 year and 2 year contract duration which has 9% and 3% churn rate respectively. 
 

In [93]:
PaperlessBilling_Churn=data2.groupby(['PaperlessBilling', 'Churn']).size()
PaperlessBilling_Churn

PaperlessBilling  Churn
No                No       2403
                  Yes       469
Yes               No       2771
                  Yes      1400
dtype: int64

People with paperless billing have higher churn rate than people with no paperless billing. 

In [94]:
PaymentMethod_Churn=data2.groupby(['PaymentMethod', 'Churn']).size()
PaymentMethod_Churn

PaymentMethod              Churn
Bank transfer (automatic)  No       1286
                           Yes       258
Credit card (automatic)    No       1290
                           Yes       232
Electronic check           No       1294
                           Yes      1071
Mailed check               No       1304
                           Yes       308
dtype: int64

In [166]:
OnlineBackup_Churn=data2.groupby(['OnlineBackup', 'Churn']).size()
OnlineBackup_Churn

OnlineBackup         Churn
No                   0        1855
                     1        1233
No internet service  0        1413
                     1         113
Yes                  0        1906
                     1         523
dtype: int64

People with no online backup have higher churn rate than people with Online backup. People with no onlien backup have 66% churn rate as compared to people with online backup having 28% churn rate. 

In [167]:
DeviceProtection_Churn=data2.groupby(['DeviceProtection', 'Churn']).size()
DeviceProtection_Churn

DeviceProtection     Churn
No                   0        1884
                     1        1211
No internet service  0        1413
                     1         113
Yes                  0        1877
                     1         545
dtype: int64

In [95]:
MonthlyCharges_Churn=data2.groupby(['MonthlyCharges', 'Churn']).size()
MonthlyCharges_Churn

MonthlyCharges  Churn
18.25           No        1
18.40           No        1
18.55           No        1
18.70           No        2
18.75           No        1
18.80           No        7
18.85           No        4
                Yes       1
18.90           No        2
18.95           No        5
                Yes       1
19.00           No        6
                Yes       1
19.05           No        9
19.10           No       11
                Yes       1
19.15           No       11
19.20           No       17
19.25           No       15
                Yes       3
19.30           No       21
                Yes       3
19.35           No       23
                Yes       2
19.40           No       27
                Yes       4
19.45           No       25
                Yes       3
19.50           No       29
                Yes       3
                         ..
115.80          No        4
115.85          No        1
116.00          No        1
116.05          No        

In [96]:
TotalCharges_Churn=data2.groupby(['TotalCharges', 'Churn']).size()
TotalCharges_Churn

TotalCharges  Churn
              No       11
100.2         No        1
100.25        Yes       1
100.35        No        1
100.4         Yes       1
100.8         Yes       1
100.9         No        1
1001.2        No        1
1001.5        No        1
1003.05       No        1
1004.35       No        1
1004.5        No        1
1004.75       No        1
1005.7        No        1
1006.9        No        1
1007.8        No        1
1007.9        No        1
1008.55       Yes       1
1008.7        No        1
              Yes       1
1009.25       No        1
101.1         No        1
101.45        Yes       1
101.65        No        1
101.9         No        1
1010          No        1
1011.05       No        1
1011.5        No        1
1011.8        No        1
1012.4        No        1
                       ..
980.35        No        1
981.45        No        1
982.95        No        1
985.05        No        1
987.95        No        1
              Yes       1
989.05        No  

In [ ]:
#Replacing the string columns with int values:
#No replaced with 0
#Yes replaced with 1
# Any other string value in the column with 2

data2['Partner'].replace(['No','Yes'],[0,1],inplace=True)
data2['Dependents'].replace(['No','Yes'],[0,1],inplace=True)
data2['PhoneService'].replace(['No','Yes'],[0,1],inplace=True)
data2['Churn'].replace(['No','Yes'],[0,1],inplace=True)
data2['PaperlessBilling'].replace(['No','Yes'],[0,1],inplace=True)
data2['MultipleLines'].replace(['No','Yes', 'No phone service'],[0,1, 2],inplace=True)
data2['OnlineSecurity'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['TechSupport'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['StreamingTV'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['StreamingMovies'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['OnlineBackup'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['DeviceProtection'].replace(['No', 'Yes', 'No internet service'],[0,1,2],inplace=True)
data2['Contract'].replace(['Month-to-month', 'One year', 'Two year'],[0,1,2],inplace=True)
data2['gender'].replace(['Male', 'Female'],[0,1],inplace=True)

In [198]:
#Look at the correlation matrix
data2.corr()

gender  SeniorCitizen   Partner  Dependents    tenure  \
gender            1.000000       0.001874  0.001808   -0.010517 -0.005106   
SeniorCitizen     0.001874       1.000000  0.016479   -0.211185  0.016567   
Partner           0.001808       0.016479  1.000000    0.452676  0.379697   
Dependents       -0.010517      -0.211185  0.452676    1.000000  0.159712   
tenure           -0.005106       0.016567  0.379697    0.159712  1.000000   
PhoneService      0.006488       0.008576  0.017706   -0.001762  0.008448   
MultipleLines     0.000485       0.099883  0.090981   -0.016875  0.242279   
OnlineSecurity    0.003429      -0.210897  0.081850    0.190523  0.145298   
OnlineBackup      0.002032      -0.152780  0.087055    0.162445  0.178651   
DeviceProtection -0.005092      -0.157095  0.094451    0.156439  0.178649   
TechSupport      -0.000985      -0.223770  0.069072    0.180832  0.144459   
StreamingTV      -0.001156      -0.130130  0.080127    0.140395  0.136145   
StreamingMovies   0.000191      -0.120802  0.075779    0.125820  0.140781   
Contract         -0.000126      -0.142554  0.294806    0.243187  0.671607   
PaperlessBilling  0.011754       0.156530 -0.014877   -0.111377  0.006152   
MonthlyCharges    0.014569       0.220173  0.096848   -0.113890  0.247900   
Churn             0.008612       0.150889 -0.150448   -0.164221 -0.352229   

                  PhoneService  MultipleLines  OnlineSecurity  OnlineBackup  \
gender                0.006488       0.000485        0.003429      0.002032   
SeniorCitizen         0.008576       0.099883       -0.210897     -0.152780   
Partner               0.017706       0.090981        0.081850      0.087055   
Dependents           -0.001762      -0.016875        0.190523      0.162445   
tenure                0.008448       0.242279        0.145298      0.178651   
PhoneService          1.000000      -0.691070        0.125353      0.150338   
MultipleLines        -0.691070       1.000000       -0.235021     -0.210372   
OnlineSecurity        0.125353      -0.235021        1.000000      0.751661   
OnlineBackup          0.150338      -0.210372        0.751661      1.000000   
DeviceProtection      0.138755      -0.200463        0.749040      0.740604   
TechSupport           0.123350      -0.232155        0.791225      0.754095   
StreamingTV           0.171538      -0.202414        0.701976      0.720671   
StreamingMovies       0.165205      -0.195815        0.704984      0.716700   
Contract              0.002247       0.078613        0.389978      0.351267   
PaperlessBilling      0.016505       0.108230       -0.334003     -0.262402   
MonthlyCharges        0.247398       0.146153       -0.621227     -0.538454   
Churn                 0.011942       0.019423       -0.332819     -0.291449   

                  DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
gender                   -0.005092    -0.000985    -0.001156         0.000191   
SeniorCitizen            -0.157095    -0.223770    -0.130130        -0.120802   
Partner                   0.094451     0.069072     0.080127         0.075779   
Dependents                0.156439     0.180832     0.140395         0.125820   
tenure                    0.178649     0.144459     0.136145         0.140781   
PhoneService              0.138755     0.123350     0.171538         0.165205   
MultipleLines            -0.200463    -0.232155    -0.202414        -0.195815   
OnlineSecurity            0.749040     0.791225     0.701976         0.704984   
OnlineBackup              0.740604     0.754095     0.720671         0.716700   
DeviceProtection          1.000000     0.767970     0.763279         0.766821   
TechSupport               0.767970     1.000000     0.737578         0.737123   
StreamingTV               0.763279     0.737578     1.000000         0.809608   
StreamingMovies           0.766821     0.737123     0.809608         1.000000   
Contract                  0.390216     0.418440     0.327951         0.33099

In [436]:
#Run Logistic regression model by including the features which had high positive and negative coorelation with churn rate
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()

feature_cols = [ 'SeniorCitizen', 'Dependents' , 'tenure',  
               'Contract', 'OnlineSecurity', 'PaperlessBilling', 'gender', 'TechSupport', 
             'OnlineBackup', 'DeviceProtection']
                
X = data2[feature_cols]
y = data2.Churn

logreg.fit(X,y)
pred = logreg.predict(X)

Features that are important predictors of churn rate are SeniorCitizen, Having dependents, gender, tenure, contract 
duration, Online security, Paperless billing, Tech support, Device Protection and Online backup. 

In [437]:
#Logistic Regression
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train.shape, X_test.shape


logreg.fit(X_train, y_train)
y_pred = logreg.predict(X_test)
print("{0} / {1} correct".format(np.sum(y_test == y_pred), len(y_test)))
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred) 

1380 / 1761 correct


0.7836456558773425

Logistic Regression model has accuracy of 79%

In [438]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.83      0.89      0.86      1304
          1       0.60      0.49      0.54       457

avg / total       0.77      0.78      0.78      1761



In [439]:
from sklearn.metrics import confusion_matrix
conf = confusion_matrix(y_test,y_pred)
print(conf)

[[1155  149]
 [ 232  225]]


Use Random Forest Model to predict churn rate

In [432]:
#Use RandomForest 
from sklearn.ensemble import RandomForestClassifier
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train.shape, X_test.shape
RF = RandomForestClassifier()
RF.fit(X_train,y_train)
predictions_RF = RF.predict(X_test)

In [433]:
#Accuracy by Random Forest
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred) 

0.6422487223168655

Random foresct model has accuracy of 64%

Use KNN Model to predict churn rate

In [375]:
#Use knn 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

In [376]:
#Accuracy using KNN
from sklearn import metrics
y_pred_class = knn.predict(X_test)
print((metrics.accuracy_score(y_test, y_pred_class)))

0.7126632595116411


KNN model has accuracy of 71%

Use Decision Tree Model to predict churn rate

In [377]:
#Decision Tree
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)

from sklearn.tree import DecisionTreeClassifier
dTree = DecisionTreeClassifier()
dTree.fit(X_train,y_train)
predictions = dTree.predict(X_test) 

In [378]:
#Accuracy using Decisiontree
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.6405451448040886

Decision Tree model has accuracy of 64%

Logistic Regression has highest accuracy of 78% and good precision and recall. 
From the 1304 data points with label = False, model was successful in identifying 1155 of those correctly and 149 were incorrectly labeled. 
From the 457 data points with label = True, model successful identified 232 of those correctly. 

Recommendations to reduce churn rate: Churn Rate can be reduced by targeting more senior citizens to sign up for services. Getting more dependents sign up for services and provide some discounts and excellent customer service so as to retain customers for long time because customers ahving high tenure have very low churn rates. Providing OnlineSecurity, PaperlessBilling, TechSupport, OnlineBackup and DeviceProtection would be good offerings to retain customers. Providing customers more discounts who are opting for one year and two year contract duration will help retain these customers as it's seen that higher the contract duration, the lesser is the churn rate. 
            